src/training/mixture/mixed_generator_dataset.py

```diff
@@ -25,6 +25,64 @@ def __init__(self, generator_population, weights, n_samples, mixture_generator_s
         weights = collections.OrderedDict(sorted(weights.items()))
         weights = {k: v for k, v in weights.items() if any([i for i in self.individuals if i.source == k])}
         weights_np = np.asarray(list(weights.values()))
+
+        if np.sum(weights_np) != 1:
+            weights_np = weights_np / np.sum(weights_np).astype(float)    # A bit of patching, but normalize it again
+
+        if mixture_generator_samples_mode == 'independent_probability':
+            self.gen_indices = np.random.choice(len(self.individuals), n_samples, p=weights_np.tolist())
+        elif mixture_generator_samples_mode == 'exact_proportion':
+            # Does not perform checking here if weights_np.tolist() sum up to one
+            # There will be some trivial error if prob*n_samples is not integer for prob in weights_np.tolist()
+            self.gen_indices = [
```

Comment:
My head hurts from reading nested list comprehensions, can we refactor it?

Reply:
This is not mine. It was there before. I have just copied it.

Reply:
So we can at least make the copied code clearer